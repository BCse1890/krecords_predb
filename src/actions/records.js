import uuid from 'uuid';
import moment from 'moment';

// ADD RECORD
export const addRecord = ( {name='', address='', contactTel='', email='', relativeName='',
    relativeContactTel='', mainDojo='', dateJoinClub=0, gradeLevel=10, 
    gradingDate=undefined, classesSinceGrading=0
    } = {} 
) => ({
    type: 'ADD_RECORD',
    record: {
        id: uuid(),
        name,
        mainDojo,
        gradeLevel,
        gradingDate,
        dateJoinClub,
        address,
        contactTel,
        email,
        relativeName,
        relativeContactTel,
        classesSinceGrading
    }
})

// DELETE RECORD
export const deleteRecord = ({ id } = {}) => ({
    type: 'DELETE_RECORD',
    id
});

// EDIT RECORD
export const editRecord = (id, updates) => ({
    type: 'EDIT_RECORD',
    id,
    updates
});