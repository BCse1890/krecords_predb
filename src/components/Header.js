import React from 'react';
import { NavLink } from 'react-router-dom';

const Header = () => (
    <header>
        <h1>Student Record Webpage</h1>
        <NavLink to="/" activeClassName="is-active" exact={true}>Homepage</NavLink>
        <NavLink to="/add" activeClassName="is-active">Create Record</NavLink>
        <NavLink to="/edit" activeClassName="is-active">Edit Record</NavLink>
        <NavLink to="/help" activeClassName="is-active">Help</NavLink>
        <NavLink to="/contact" activeClassName="is-active">Registration details</NavLink> 
    </header>
);

export default Header;