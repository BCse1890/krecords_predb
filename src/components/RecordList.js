import React from 'react';
import { connect } from 'react-redux';
import RecordListItem from './RecordListItem';
import selectRecords from '../selectors/records';


// props passed through from ConnectedRecordList below, drawing
// state.records prop from store
export const RecordList = (props) => (
    <div>
        {
            props.records.length === 0 ? (
                <p>No expenses</p>
            ) : (
                props.records.map((record) => {
                    return <RecordListItem key={record.id} {...record} />;
                })
            )
        }
        
    </div>
);

// connects to store and creates state.records as prop passed
// to RecordList function above

/* const ConnectedRecordList = connect((state) => {
    return {
        records: state.records
    }

})(RecordList);

export default ConnectedRecordList; */

// common not to create separate ConnectedRecordList variable but
// manage as an export default 
/* export default  connect((state) => {
    return {
        records: state.records
    }

})(RecordList); */


// we can pull out the state to props code and create const mapStateToProps
const mapStateToProps = (state) => {
    return {
        records: selectRecords(state.records, state.filters)
        // records: state.records,
        // filters: state.filters
    };
};

export default connect(mapStateToProps)(RecordList);