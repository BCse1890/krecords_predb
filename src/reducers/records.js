
// create kRecords Reducer
const recordsReducerDefault = [];

const recordsReducer = (state = recordsReducerDefault, action) => {
    switch(action.type) {
        case 'ADD_RECORD':
            return [
                //spreads out all existing records in array
                // and adds on a new array through action.record
                ...state,
                action.record
            ];   
        case 'DELETE_RECORD':
            return state.filter(({ id }) => (id !== action.id) );
        case 'EDIT_RECORD':
            return state.map((record) => {
                    if(record.id === action.id) {
                        return {
                            ...record,
                            ...action.updates
                        }
                    } else {
                        return record;
                    }
                });
        default:
            return state;
    }
};

export default recordsReducer;