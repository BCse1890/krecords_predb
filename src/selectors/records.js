import moment from 'moment';

// Get Visible Expenses
export default (records, { text, sortBy, startDate, endDate }) => {
//const getVisibleExpenses = (records, { text, gradingDate, sortBy, startDate, endDate }) => {
    return records.filter((record) => {
        // filter out if startDate not a number or record.created >= startDate;
        const gradeDateMoment = moment(record.gradingDate);
        const startDateMatch = startDate ? startDate.isSameOrBefore(gradeDateMoment, 'day') : true;
        const endDateMatch = endDate ? endDate.isSameOrAfter(gradeDateMoment, 'day') : true;
        
        const textNameMatch = record.name.toLowerCase().includes(text.toLowerCase());
        const textDojoMatch = record.mainDojo.toLowerCase().includes(text.toLowerCase());
        
            
        return startDateMatch && endDateMatch && textDojoMatch || textNameMatch;
    }).sort((a, b) => {
        if(sortBy === 'mainDojo') {
            return a.mainDojo < b.mainDojo ? -1 : 1;
        } else if(sortBy === 'gradeLevel') {
            return a.gradeLevel < b.gradeLevel ? 1 : -1;
        }
    });
};

// export default getVisibleExpenses;