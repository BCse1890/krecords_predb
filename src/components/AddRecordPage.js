import React from 'react';
import { connect } from 'react-redux';
import RecordForm from './RecordForm';
import { addRecord } from '../actions/records';

export class AddRecordPage extends React.Component {
    onSubmit = (record) => {
        this.props.addRecord(record);
        this.props.history.push('/');
    };
    render() {
        return (
            <div>
            <h1>Add Record</h1>
            <RecordForm 
              onSubmit={this.onSubmit}
            />
        </div>
        );
    }
}

const mapDispatchToProps = (dispatch) => ({
        addRecord: (record) => dispatch(addRecord(record))
    });

export default connect(undefined, mapDispatchToProps)(AddRecordPage);