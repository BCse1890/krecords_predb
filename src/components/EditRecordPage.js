import React from 'react';
import { connect } from 'react-redux';
import { Link } from 'react-router-dom';
import RecordForm from './RecordForm';
import { editRecord, deleteRecord } from '../actions/records';

export class EditRecordPage extends React.Component {
    onSubmit = (record) => {
        this.props.editRecord(this.props.record.id, record);
        this.props.history.push('/');
    };
    onDelete = () => {
        console.log("Delete key pressed");
        this.props.deleteRecord({ id: this.props.record.id });
        this.props.history.push('/');
    };
    render() {
        return (
            <div>
                <RecordForm 
                    record={this.props.record}
                    onSubmit={this.onSubmit}
                />
                <button onClick={this.onDelete}>Delete</button>
                <br/>
            </div>
        );
    }
};

const mapStateToProps = (state, props) => ({
    record: state.records.find((record) => 
        record.id === props.match.params.id)
});

const mapDispatchToProps = (dispatch, props) => ({
    editRecord: (id, record) =>  dispatch(editRecord(id, record)),
    deleteRecord: (data) => dispatch(deleteRecord(data))
});

export default connect (mapStateToProps, mapDispatchToProps)(EditRecordPage);