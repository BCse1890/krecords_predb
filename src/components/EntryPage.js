import React from 'react';
import RecordList from './RecordList';
import RecordListFilters from './RecordListFilters';

const EntryPage = () => (
    <div>
        <RecordListFilters />
        <RecordList/>
    </div>
);


export default EntryPage;