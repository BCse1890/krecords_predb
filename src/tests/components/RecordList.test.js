import React from 'react';
import { shallow, mount, render } from 'enzyme';
import '../setUpTests';
import { RecordList } from '../../components/RecordList';
import records from '../fixtures/records';

test('should render RecordList with records', () => {
    const wrapper = shallow(<RecordList records={records}/>);
    expect(wrapper).toMatchSnapshot();
});

test('shoulder render RecordList with empty message', () => {
    const wrapper = shallow(<RecordList records={[]} />);
    expect(wrapper).toMatchSnapshot();
});

