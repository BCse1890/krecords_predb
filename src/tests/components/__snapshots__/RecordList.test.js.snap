// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render RecordList with records 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <RecordList
    records={
      Array [
        Object {
          "gradeLevel": 3,
          "gradingDate": 0,
          "id": "1",
          "mainDojo": "Keynsham",
          "name": "Bill Jo",
        },
        Object {
          "gradeLevel": 4,
          "gradingDate": 2,
          "id": "2",
          "mainDojo": "Emersons Green",
          "name": "Bo Ko",
        },
        Object {
          "gradeLevel": 8,
          "gradingDate": 24,
          "id": "3",
          "mainDojo": "Filton",
          "name": "See Yay",
        },
      ]
    }
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": Array [
        <RecordListItem
          gradeLevel={3}
          gradingDate={0}
          id="1"
          mainDojo="Keynsham"
          name="Bill Jo"
        />,
        <RecordListItem
          gradeLevel={4}
          gradingDate={2}
          id="2"
          mainDojo="Emersons Green"
          name="Bo Ko"
        />,
        <RecordListItem
          gradeLevel={8}
          gradingDate={24}
          id="3"
          mainDojo="Filton"
          name="See Yay"
        />,
      ],
    },
    "ref": null,
    "rendered": Array [
      Object {
        "instance": null,
        "key": "1",
        "nodeType": "function",
        "props": Object {
          "gradeLevel": 3,
          "gradingDate": 0,
          "id": "1",
          "mainDojo": "Keynsham",
          "name": "Bill Jo",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "2",
        "nodeType": "function",
        "props": Object {
          "gradeLevel": 4,
          "gradingDate": 2,
          "id": "2",
          "mainDojo": "Emersons Green",
          "name": "Bo Ko",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
      Object {
        "instance": null,
        "key": "3",
        "nodeType": "function",
        "props": Object {
          "gradeLevel": 8,
          "gradingDate": 24,
          "id": "3",
          "mainDojo": "Filton",
          "name": "See Yay",
        },
        "ref": null,
        "rendered": null,
        "type": [Function],
      },
    ],
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": Array [
          <RecordListItem
            gradeLevel={3}
            gradingDate={0}
            id="1"
            mainDojo="Keynsham"
            name="Bill Jo"
          />,
          <RecordListItem
            gradeLevel={4}
            gradingDate={2}
            id="2"
            mainDojo="Emersons Green"
            name="Bo Ko"
          />,
          <RecordListItem
            gradeLevel={8}
            gradingDate={24}
            id="3"
            mainDojo="Filton"
            name="See Yay"
          />,
        ],
      },
      "ref": null,
      "rendered": Array [
        Object {
          "instance": null,
          "key": "1",
          "nodeType": "function",
          "props": Object {
            "gradeLevel": 3,
            "gradingDate": 0,
            "id": "1",
            "mainDojo": "Keynsham",
            "name": "Bill Jo",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "2",
          "nodeType": "function",
          "props": Object {
            "gradeLevel": 4,
            "gradingDate": 2,
            "id": "2",
            "mainDojo": "Emersons Green",
            "name": "Bo Ko",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
        Object {
          "instance": null,
          "key": "3",
          "nodeType": "function",
          "props": Object {
            "gradeLevel": 8,
            "gradingDate": 24,
            "id": "3",
            "mainDojo": "Filton",
            "name": "See Yay",
          },
          "ref": null,
          "rendered": null,
          "type": [Function],
        },
      ],
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;

exports[`shoulder render RecordList with empty message 1`] = `
ShallowWrapper {
  Symbol(enzyme.__root__): [Circular],
  Symbol(enzyme.__unrendered__): <RecordList
    records={Array []}
  />,
  Symbol(enzyme.__renderer__): Object {
    "batchedUpdates": [Function],
    "getNode": [Function],
    "render": [Function],
    "simulateError": [Function],
    "simulateEvent": [Function],
    "unmount": [Function],
  },
  Symbol(enzyme.__node__): Object {
    "instance": null,
    "key": undefined,
    "nodeType": "host",
    "props": Object {
      "children": <p>
        No expenses
      </p>,
    },
    "ref": null,
    "rendered": Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": "No expenses",
      },
      "ref": null,
      "rendered": "No expenses",
      "type": "p",
    },
    "type": "div",
  },
  Symbol(enzyme.__nodes__): Array [
    Object {
      "instance": null,
      "key": undefined,
      "nodeType": "host",
      "props": Object {
        "children": <p>
          No expenses
        </p>,
      },
      "ref": null,
      "rendered": Object {
        "instance": null,
        "key": undefined,
        "nodeType": "host",
        "props": Object {
          "children": "No expenses",
        },
        "ref": null,
        "rendered": "No expenses",
        "type": "p",
      },
      "type": "div",
    },
  ],
  Symbol(enzyme.__options__): Object {
    "adapter": ReactSixteenAdapter {
      "options": Object {
        "enableComponentDidUpdateOnSetState": true,
        "lifecycles": Object {
          "componentDidUpdate": Object {
            "onSetState": true,
          },
          "getDerivedStateFromProps": true,
          "getSnapshotBeforeUpdate": true,
          "setState": Object {
            "skipsComponentDidUpdateOnNullish": true,
          },
        },
      },
    },
    "attachTo": undefined,
    "hydrateIn": undefined,
  },
}
`;
